File: shoe/src/main/java/com/dan/shoe/shoe/services/OrderService.java
package com.dan.shoe.shoe.services;

import com.dan.shoe.shoe.dtos.requests.OrderCreationByStaff;
import com.dan.shoe.shoe.dtos.requests.OrderNowCreation;
import com.dan.shoe.shoe.dtos.requests.OrderProductCreation;
import com.dan.shoe.shoe.dtos.responses.OrderStats;
import com.dan.shoe.shoe.dtos.responses.ResponseMessage;
import com.dan.shoe.shoe.dtos.responses.Statuses;
import com.dan.shoe.shoe.models.Cart;
import com.dan.shoe.shoe.models.Order;
import com.dan.shoe.shoe.models.User;
import com.dan.shoe.shoe.models.enums.OrderStatus;
import com.dan.shoe.shoe.models.enums.OrderType;
import com.dan.shoe.shoe.models.enums.PaymentType;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;

import java.time.LocalDate;
import java.util.List;
import java.util.Map;

public interface OrderService {
    Order createOrder(String username, String voucherCode, PaymentType paymentType, String address);
    Order getOrderById(Long id);
    Page<Order> getOrdersByUser(Pageable pageable, String username);
    void updateOrderStatus(Long orderId, String status);
    void updateOrderPaid(Long orderId);
    Page<Order> getAllOrders(Pageable pageable, String keyword, LocalDate startDate, LocalDate endDate);
    Order createOrderByStaff(String username, OrderCreationByStaff orderCreationByStaff);
    Order getOrderInfo(Long orderId);
    Order createOrderNow(String username, OrderNowCreation orderNowCreation);
    Order switchOrderStatus(Long orderId, String username);
    Page<Order> getOrderByOrderType(Pageable pageable, OrderType orderType);
    Page<Order> getOrderByStatus(OrderStatus status, String keyword, LocalDate startDate, LocalDate endDate, Pageable pageable);
    ResponseMessage deleteOrder(Long orderId);
    List<Statuses> getOrderStatuses();
    List<Statuses> getOrderStatusesForUser(String username);
    Page<Order> getOrdersByUserAndStatus(String username, OrderStatus status, Pageable pageable);
    ResponseMessage cancelOrder(Long orderId);
    Map<String, Object> getRevenueAndOrderData();
    Map<String, Object> getDailyRevenueData(LocalDate startDate, LocalDate endDate);
}

File: shoe-app/src/services/order.service.ts
import axios from 'axios';
import { OrderCreationByStaff } from '../models/request/OrderCreationByStaff';
import { NowCreation } from '../models/request/NowCreation';

const BASE_URL = process.env.REACT_APP_BASE_URL

export const createOrder = async (voucherCode: string, paymentType: string, address: string) => {
    try {
        const response = await axios.post(`${BASE_URL}/orders/create?voucherCode=${voucherCode}&paymentType=${paymentType}&address=${address}`,
            {},
            {
                headers: {
                    Authorization: `Bearer ${localStorage.getItem('accessToken')}`
                }
            }
        );
        return response.data;
    } catch (error) {
        console.error('Error during create account:', error);
        throw error;
    }
};

export const createOrderByStaff = async (orderCreationByStaff: OrderCreationByStaff) => {
    try {
        const response = await axios.post(`${BASE_URL}/orders/staff/create`, 
            orderCreationByStaff,
            {
                headers: {
                    Authorization: `Bearer ${localStorage.getItem('accessToken')}`
                }
            }
        );
        return response.data;
    } catch (error) {
        console.error('Error during create account:', error);
        throw error;
    }
};

export const getOrdersByStatus = async (keyword = '', page = 0, size = 10, sortBy = '', order = '', status: string, startDate: null, endDate: null) => {
    try {
        const response = await axios.get(`${BASE_URL}/orders/get-by-status`, {
            params: {
                keyword,
                page,
                size,
                sortBy,
                order,
                status,
                startDate,
                endDate
            },
            headers: {
                Authorization: `Bearer ${localStorage.getItem('accessToken')}`
            }
        });
        return response;
    } catch (error) {
        console.error('Error during get order infor:', error);
        throw error;
    }
}

export const createOrderNow = async (nowCreation: NowCreation) => {
    try {
        const response = await axios.post(`${BASE_URL}/orders/now`,
            nowCreation,
            {
                headers: {
                    Authorization: `Bearer ${localStorage.getItem('accessToken')}`
                }
            }
        );
        return response.data;
    } catch (error) {
        console.error('Error during create account:', error);
        throw error;
    }
};

export const getOrderInfor = async (id: number) => {
    try {
        const response = await axios.get(`${BASE_URL}/orders/info/${id}`, {
            headers: {
                Authorization: `Bearer ${localStorage.getItem('accessToken')}`
            }
        });
        return response;
    } catch (error) {
        console.error('Error during get order infor:', error);
        throw error;
    }
}

export const deleteOrder = async (id: number) => {
    try {
        const response = await axios.delete(`${BASE_URL}/orders/admin/delete/${id}`, {
            headers: {
                Authorization: `Bearer ${localStorage.getItem('accessToken')}`
            }
        });
        return response.data;
    } catch (error) {
        console.error('Error during delete order:', error);
        throw error;
    }
}

export const getOrderStatuses = async () => {
    try {
        const response = await axios.get(`${BASE_URL}/orders/admin/get/statuses`, {
            headers: {
                Authorization: `Bearer ${localStorage.getItem('accessToken')}`
            }
        });
        return response.data;
    } catch (error) {
        console.error('Error during get order statuses:', error);
        throw error;
    }
}

export const getOrderStatusesForUser = async () => {
    try {
        const response = await axios.get(`${BASE_URL}/orders/user/get-statuses`, {
            headers: {
                Authorization: `Bearer ${localStorage.getItem('accessToken')}`
            }
        });
        return response.data;
    } catch (error) {
        console.error('Error during get order statuses:', error);
        throw error;
    }
}

export const cancelOrder = async (id: number) => {
    try {
        const response = await axios.put(`${BASE_URL}/orders/public/cancel/${id}`, {},
            {
                headers: {
                    Authorization: `Bearer ${localStorage.getItem('accessToken')}`
                }
            }
        );
        return response.data;
    } catch (error) {
        console.error('Error during cancel order:', error);
        throw error;
    }
}

export const updateOrderStatus = async (id: number) => {
    try {
        const response = await axios.put(`${BASE_URL}/orders/staff/switch-status/${id}`, 
            {},
            {
                headers: {
                    Authorization: `Bearer ${localStorage.getItem('accessToken')}`
                }
            }
        );
        return response.data;
    } catch (error) {
        console.error('Error during update order status:', error);
        throw error;
    }
}

export const getMyOrders = async (page: number, status: string) => {
    try {
        const response = await axios.get(`${BASE_URL}/orders/my-orders?page=${page}&status=${status}`, {
            headers: {
                Authorization: `Bearer ${localStorage.getItem('accessToken')}`
            }
        });
        return response;
    } catch (error) {
        console.error('Error during get my orders:', error);
        throw error;
    }
}

export const getStatistics = async () => {
    try {
        const response = await axios.get(`${BASE_URL}/orders/statistics`, {
            headers: {
                Authorization: `Bearer ${localStorage.getItem('accessToken')}`
            }
        });
        return response;
    } catch (error) {
        console.error('Error during get statistics:', error);
        throw error;
    }
}

export const getDailyStatistics = async ( startDate: string, endDate: string ) => {
    return axios.get(`${BASE_URL}/orders/daily-statistics`, {
        params: {
            startDate,
            endDate
        },
        headers: {
            Authorization: `Bearer ${localStorage.getItem('accessToken')}`
        }
    });
};

export const updateOrderPaid = async (id: number) => {
    return axios.put(`${BASE_URL}/orders/public/update-paid/${id}`, {}, {
        headers: {
            Authorization: `Bearer ${localStorage.getItem('accessToken')}`
        }
    });
}
